### PR TITLE
refactor(crm): v2 E-company-deal-task

### PR DESCRIPTION
**Pull request by dulsara-manakal**
### Summary

-

### How to test

1.

### Project Checklist

- [ ] Changes build without any errors
- [ ] Have written adequate test cases
- [ ] Done developer testing in
  - [ ] Chrome
  - [ ] Firefox
  - [ ] Safari
- [ ] Code is formatted with `npm run format`
- [ ] Code is linted with `npm run check-lint`
- [ ] No unnecessary comments left in code
- [ ] Made corresponding changes to the documentation

#### Other

- [ ] New atomic components added
- [ ] New molecules added
- [ ] New pages(routes) added
- [ ] New dependencies installed

### PR Checklist

- [ ] Pull request is raised from the correct source branch
- [ ] Pull request is raised to the correct destination branch
- [ ] Pull request is raised with correct title
- [ ] Pull request is self reviewed
- [ ] Pull request is self assigned
- [ ] Suitable pull request status labels are added (`ready-for-code-review`)

### Additional Information
-

### CODE REVIEW

**Review by ThinuwanW — COMMENTED** [review]
## 🤖 Claude Code Review

This PR adds the v2 CRM contact modal stack (Add/Edit/controller, shared form, company + owner editable fields, email-duplicate hook), wires the v2 contacts page behind a `false` feature flag, expands CompanySidePanel to load tasks/deals/contacts, and migrates DealSidePanelV2 onto the v2 task API + normalized store. The new code follows the established v2 patterns closely (store-as-normalized-cache, Formik+Yup, translation keys all present and verified). The material risks are on DealSidePanelV2 — which is live today since `isCrmDealsV2 = true` — where task ids are now derived from query data while the task entities land in the store only via a post-paint effect, and the task query lost its `enabled` guard; plus a missing TaskModalController on the new v2 contacts page and a few store/effect dependency issues.

Found **13** new issue(s): 🔴 3 important, 🟡 7 suggestion(s), 🟣 3 nit(s)

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx [hunk 1]
```diff
@@ -102,16 +102,24 @@ const DealSidePanelV2: FC = () => {
...
     isFetchingNextPage
-  } = useGetRelatedTasks({ dealId: selectedDealId, size: TASK_PAGE_SIZE });
+  } = useGetTasksInfinite({
+    dealId: selectedDealId ?? undefined,
+    size: TASK_PAGE_SIZE
+  });
+
+  const taskItems = fetchedTasks?.pages.flatMap((page) => page.items) ?? [];
+
+  const taskIds = toTaskIds(taskItems);
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: 🤖 **Claude** · 🔴 **Important**

`taskIds` is computed at render from the query response, but `SidePanelTasksSection` resolves those ids against the Zustand `tasks` record (`resolveTasks(taskIds, optimisticTasks)`), which is only populated by the `useEffect` on line 118. `useEffect` runs after paint, so on the first render after `fetchedTasks` arrives the section takes the `taskIds?.length` branch and renders `SidePanelTasksList` with an empty array — a visible flash of an empty task list before the effect commits the store write. The previous code passed `relatedTasks` straight through, so there was no gap. Note `CompanySidePanel` doesn't have this problem because it derives `taskIds` from `company.taskIds`, which is written in the same effect/batch as `setTasks`. Fix: either resolve from the fetched items at render (`updateTaskRecord(tasks, taskItems)` computed in a `useMemo` and passed down), or keep the store as the single source and read the ids back from it rather than from the query.

File: frontend/pages/community/crm/contacts.tsx [hunk 2]
```diff
@@ -64,4 +74,57 @@ const Contacts: NextPage = () => {
...
+      module={Modules.CRM}
+    >
+      <>
+        {selectedContactId && (
+          <SidePanelWrapperV2>
+            <ContactSidePanelV2 contactId={selectedContactId} />
+          </SidePanelWrapperV2>
+        )}
+
+        <ContactModalControllerV2 />
```
Comment (ThinuwanW) [inline frontend/pages/community/crm/contacts.tsx:121]: 🤖 **Claude** · 🔴 **Important**

`ContactsV2` drops the `<TaskModalController />` that `ContactsV1` mounts (line 70) and does not mount a v2 equivalent. `ContactSidePanelV2` renders `SidePanelTasksSection` with `showAddTaskAction` defaulting to true, and its `handleAddTask` calls `setTaskModalType(CrmModalTypes.ADD_TASK_MODAL)` / `setIsTaskModalOpen(true)` on the v2 store — with no controller mounted, both the toolbar 'Add task' button and the empty-state 'Add task' button are silent no-ops. `TaskModalControllerV2` exists (`~community/crm/v2/components/organisms/TaskModalController/TaskModalController`) but is currently only mounted on `pages/community/crm/tasks.tsx`. Add `<TaskModalControllerV2 />` alongside `<ContactModalControllerV2 />`, or pass `showAddTaskAction={false}` down. (The same gap already exists on the v2 deals/companies pages, so this is worth fixing there too — deals v2 is live.)

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx [hunk 3]
```diff
@@ -102,16 +102,24 @@ const DealSidePanelV2: FC = () => {
   };
 
   const {
-    data: relatedTasksData,
+    data: fetchedTasks,
     fetchNextPage,
     hasNextPage,
     isFetchingNextPage
-  } = useGetRelatedTasks({ dealId: selectedDealId, size: TASK_PAGE_SIZE });
+  } = useGetTasksInfinite({
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: 🤖 **Claude** · 🔴 **Important**

`useGetTasksInfinite` is called without the `enabled` argument it accepts (`TaskApi.ts:43`), and the filter is `{ dealId: selectedDealId ?? undefined, size: TASK_PAGE_SIZE }`. If `selectedDealId` is ever null, `dealId` drops out of the params and the query fetches the entire unfiltered org task list, which the effect on line 118 then dumps into the shared `tasks` store. The only thing preventing this today is the `{selectedDealId !== null && <DealSidePanelV2 />}` guard in `pages/community/crm/deals.tsx:104` — a guard in a different file, on a page that is live (`isCrmDealsV2 = true`). The old `useGetRelatedTasks` call was similarly unguarded but hit a deal-scoped endpoint, so this is a new blast radius. Pass `selectedDealId != null` as the second argument, matching how every other v2 call site gates its lookups.

File: frontend/src/community/crm/v2/components/organisms/CompanySidePanel/CompanySidePanel.tsx [hunk 4]
```diff
@@ -40,62 +70,184 @@ interface CompanySidePanelProps {
...
+    if (fetchedContacts) {
+      const contactItems = fetchedContacts.pages.flatMap(
+        (page) => page.items ?? []
+      );
+
+      setContacts(updateContactRecord(contacts, contactItems));
+      companyFields.contactIds = toContactIds(contactItems);
+    }
+
+    setCompanies(updateCompany(companies, companyId, companyFields));
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/organisms/CompanySidePanel/CompanySidePanel.tsx:228]: 🤖 **Claude** · 🟡 **Suggestion**

This effect calls `setTasks`/`setDeals`/`setContacts`/`setCompanies` using the `tasks`, `deals`, `contacts` and `companies` values captured at render, but none of them appear in the dependency array (only the query results and `companyId`). Two concrete consequences: (1) `setCompanies(updateCompany(companies, companyId, companyFields))` runs unconditionally, so on the very first mount — before any query resolves — it writes `companies[companyId] = {}`. That empty stub is truthy, so `getMissingCompanyIds` (companyUtil.ts:119), which tests `!companies[id]`, will treat that company as already loaded and other screens (e.g. `ContactTable`) will never backfill its name. (2) Any store write made between effect runs by another subscriber is read from a snapshot rather than current state. Guard the write with `if (Object.keys(companyFields).length > 0)` and, where possible, use functional updates so each setter reads current state rather than a captured snapshot.

File: frontend/src/community/crm/v2/components/molecules/EditableContactOwnerField/EditableContactOwnerField.tsx [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  );
+
+  const { owners, setOwners } = useCrmStoreV2(
+    useShallow((state) => ({
+      owners: state.owners,
+      setOwners: state.setOwners
+    }))
+  );
+
+  const { data: ownerLookupData, isFetching } = useGetOwnerLookupV2(
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/molecules/EditableContactOwnerField/EditableContactOwnerField.tsx:49]: 🤖 **Claude** · 🟡 **Suggestion**

`useGetOwnerLookupV2(debouncedOwnerSearch, DEFAULT_LOOKUP_PAGE_SIZE, true)` hardcodes `enabled: true`. Because `ContactModalController` renders `getModalContent()` regardless of `isOpen`, this component is mounted for the whole life of the contacts page, and hooks run before the `if (selectedOwner) return <SelectedOwnerField .../>` early return on line 40. So the owner-lookup request fires on every page load even when the modal is closed and an owner is already selected (which is the default in the Add flow, since `ownerId` is seeded from the current user). The established v2 pattern gates it — see `OwnerPopupSearch.tsx:49` (`isCrmSalesManager ?? false`). Move the `getOwnerById` call above the query and pass `!selectedOwner` (or accept an `isEnabled` prop from `ContactModalForm`).

File: frontend/src/community/crm/v2/components/molecules/ContactModalForm/ContactModalForm.tsx [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+      {canEditOwner ? (
+        <EditableContactOwnerField
+          ownerId={values.ownerId}
+          errorMessage={touched.ownerId ? errors.ownerId : undefined}
+          translateText={translateText}
+          onChange={(owner) => setFieldValue("ownerId", owner?.employeeId)}
+        />
+      ) : (
+        selectedOwner && (
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/molecules/ContactModalForm/ContactModalForm.tsx:182]: 🤖 **Claude** · 🟡 **Suggestion**

When `canEditOwner` is false the owner is rendered only if `getOwnerById(owners, values.ownerId)` resolves from the Zustand `owners` record. The v1 form took an explicit `initialOwner` prop built from `useGetUserPersonalDetails` (see `crm/components/molecules/AddContactModalContent/AddContactModalContent.tsx:61`), so it always had something to render. In v2, if the seeded owner isn't in `owners` (the record is populated from the board-init payload and from lookup results only), a non-sales-manager sees no owner field at all — and since `ownerId` is `.required()` in `getContactValidationSchema` (contactValidations.ts:42), `errors.ownerId` is only surfaced in the `EditableContactOwnerField` branch, so a failing submit would show no error anywhere. Render a fallback read-only owner from the current-user data, and surface `errors.ownerId` in the non-editable branch too.

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx [hunk 7]
```diff
@@ -102,16 +102,24 @@ const DealSidePanelV2: FC = () => {
...
+
+  const taskItems = fetchedTasks?.pages.flatMap((page) => page.items) ?? [];
+
+  const taskIds = toTaskIds(taskItems);
 
-  const relatedTasks =
-    (
-      relatedTasksData as unknown as InfiniteData<RelatedTasksPage> | undefined
-    )?.pages.flatMap((page) => page.items) ?? [];
+  useEffect(() => {
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: 🤖 **Claude** · 🟡 **Suggestion**

This effect merges every cached page back into the store whenever `fetchedTasks` changes, and `updateTaskRecord` lets the incoming server object win (`{...existing[id], ...task}`). Combined with `SidePanelTasksSection`'s completion toggle — which writes the new `isCompleted` straight to the store on mutation success while `useUpdateTask` (TaskApi.ts:189) invalidates only the company/contact METRICS roots, never `crmTaskQueryKeys.LIST` — a user who ticks a task complete and then scrolls to load the next page will see the tick revert, because page 0 is still cached with the pre-toggle value and gets re-merged. Either add `queryClient.invalidateQueries({ queryKey: crmTaskQueryKeys.LIST(...) })` to `useUpdateTask`, or make the merge non-clobbering for locally-updated fields.

File: frontend/src/community/crm/v2/components/molecules/AddContactModalContent/AddContactModalContent.tsx [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+      setContacts: state.setContacts,
+      setContactIds: state.setContactIds,
+      setCompanies: state.setCompanies,
+      setIsContactModalOpen: state.setIsContactModalOpen
+    }))
+  );
+
+  const { data: currentUser } = useGetUserPersonalDetails();
+
+  const formik = useFormik<CrmContactEntity>({
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/molecules/AddContactModalContent/AddContactModalContent.tsx:47]: 🤖 **Claude** · 🟡 **Suggestion**

`enableReinitialize: true` combined with `ownerId: currentUser?.employeeId ? Number(...) : undefined` means the form resets when `useGetUserPersonalDetails` resolves: Formik deep-compares `initialValues`, and `ownerId` genuinely changes from `undefined` to a number, so `resetForm()` fires and discards anything the user typed in the meantime. The `['me']` query is usually warm, but the modal can be opened on a cold cache. The same class of issue exists in `EditContactModalContent` (line 49), where `initialValues` is memoised off `selectedContact` — i.e. off the store `contacts` record that the open side panel's `useGetContactById` keeps refreshing, so a background refetch that actually changes a field will wipe in-progress edits. Prefer dropping `enableReinitialize` and seeding `ownerId` via `setFieldValue` only while it's still empty.

File: frontend/src/community/crm/v2/components/molecules/EditContactModalContent/EditContactModalContent.tsx [hunk 9]
```diff
@@ -0,0 +1,150 @@
...
+      setSubmitting(false);
+      return;
+    }
+
+    const changedFields = getChangedContactFields(initialValues, {
+      name: values.name?.trim(),
+      email: values.email?.trim(),
+      contactNumber: values.contactNumber?.trim(),
+      companyId: values.companyId,
+      companyName: values.companyName?.trim(),
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/molecules/EditContactModalContent/EditContactModalContent.tsx:122]: 🤖 **Claude** · 🟡 **Suggestion**

`companyName` is passed into `getChangedContactFields`, but that helper (contactUtil.ts:248) only diffs `name`, `email`, `contactNumber`, `companyId` and `ownerId` — `companyName` is silently dropped from the PATCH payload. It's harmless today because `canAddNewCompany` isn't passed to `ContactModalForm` in the edit flow, so `onAddNew` can never fire and `companyName` is always undefined. But it reads as supported and would fail silently the moment edit gains the add-new-company affordance. Either remove the field from the call or add a `companyName` branch to `getChangedContactFields`.

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx [hunk 10]
```diff
@@ -68,16 +68,16 @@ const DealSidePanelV2: FC = () => {
...
 
-  const { data: dealDetail, isFetchedAfterMount } = useGetDealById(
+  const { data: dealDetail } = useGetDealById(
     selectedDealId ?? 0,
     selectedDealId != null
   );
 
   useEffect(() => {
-    if (dealDetail && isFetchedAfterMount) {
+    if (dealDetail) {
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: 🤖 **Claude** · 🟡 **Suggestion**

The `isFetchedAfterMount` guard was dropped from this effect (`if (dealDetail && isFetchedAfterMount)` -> `if (dealDetail)`). That guard specifically prevented merging a cached — possibly stale — `dealDetail` into the store on mount, which matters here because `handleSuccess`/`ingestEditedDeal` may have already written fresher values from an inline edit. Please confirm this was intentional rather than incidental cleanup. Separately, the effect reads `deals` for `mergeDeals(deals, [dealDetail])` but `deals` isn't in the dependency array.

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx [hunk 11]
```diff
@@ -102,16 +102,24 @@ const DealSidePanelV2: FC = () => {
...
     fetchNextPage,
     hasNextPage,
     isFetchingNextPage
-  } = useGetRelatedTasks({ dealId: selectedDealId, size: TASK_PAGE_SIZE });
+  } = useGetTasksInfinite({
+    dealId: selectedDealId ?? undefined,
+    size: TASK_PAGE_SIZE
+  });
+
+  const taskItems = fetchedTasks?.pages.flatMap((page) => page.items) ?? [];
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: 🤖 **Claude** · 🟣 **Nit**

Inconsistent defensive handling of the paged response across the three call sites touched here: this line is `pages.flatMap((page) => page.items)`, `CompanySidePanel.tsx:204` is `page.items ?? []`, and the existing `ContactTable.tsx:111` is `page?.items ?? []`. `fetchTasks` returns `response?.data?.results?.[0]` (TaskApi.ts:40), which can be `undefined` on an unexpected payload — in which case this line throws a TypeError rather than degrading. Standardise on `page?.items ?? []`.

File: frontend/src/community/crm/v2/components/molecules/EditableContactCompanyField/EditableContactCompanyField.tsx [hunk 12]
```diff
@@ -0,0 +1,209 @@
...
+  useSearchCompaniesByDomain
+} from "~community/crm/v2/api/CompanyApi";
+import {
+  DEFAULT_LOOKUP_PAGE_SIZE,
+  DOMAIN_SEARCH_LIMIT
+} from "~community/crm/v2/constants/commonConstants";
+import { ADD_NEW_COMPANY_OPTION_ID } from "~community/crm/v2/constants/contactConstants";
+import { useCrmStoreV2 } from "~community/crm/v2/store/store";
+import { CrmCompanyEntity } from "~community/crm/v2/types/CrmCommonTypes";
+import { CrmCompanyDomainSearchFilterRequest } from "~community/crm/v2/types/CrmTypes";
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/components/molecules/EditableContactCompanyField/EditableContactCompanyField.tsx]: 🤖 **Claude** · 🟣 **Nit**

Two consecutive import statements pull from the same module: `CrmCompanyDomainSearchFilterRequest` and `CrmCompanyFilterRequest` are both from `~community/crm/v2/types/CrmTypes`. Merge them into one import.

File: frontend/src/community/crm/v2/hooks/useContactEmailCheck.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
+import { SEARCH_DEBOUNCE_DELAY } from "~community/common/constants/commonConstants";
```
Comment (ThinuwanW) [inline frontend/src/community/crm/v2/hooks/useContactEmailCheck.ts]: 🤖 **Claude** · 🟣 **Nit**

This new hook carries all the duplicate-email gating logic for both the add and edit flows (it is what disables the Save button via `isEmailCheckUnresolved`/`isDuplicateEmail`), and the debounce/validity/original-email interaction has several non-obvious edge cases — e.g. an email-check request that errors leaves `emailExistsData` undefined and therefore silently allows submit. The sibling hook `useInitializeCrmData` has a `.test.ts`; consider adding one here covering: unchanged email (check disabled), invalid email (check disabled), in-flight debounce (unresolved), and duplicate found.

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx (hunk 1 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: Gone. The deal panel was split on develop, this code no longer exists.

File: frontend/pages/community/crm/contacts.tsx (hunk 2 above)
Comment (dulsara-manakal) [reply frontend/pages/community/crm/contacts.tsx:121]: Fixed. Mounted TaskModalControllerV2.

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx (hunk 3 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: Gone. develop went back to useGetRelatedTasks, and the null guard is in the component now.

File: frontend/src/community/crm/v2/components/organisms/CompanySidePanel/CompanySidePanel.tsx (hunk 4 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/organisms/CompanySidePanel/CompanySidePanel.tsx:228]: Fixed. Skipping the write when there is nothing to merge.

File: frontend/src/community/crm/v2/components/molecules/EditableContactOwnerField/EditableContactOwnerField.tsx (hunk 5 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/molecules/EditableContactOwnerField/EditableContactOwnerField.tsx:49]: Invalid. BasicModal returns null when closed, so the modal content never mounts and the lookup does not fire.

File: frontend/src/community/crm/v2/components/molecules/ContactModalForm/ContactModalForm.tsx (hunk 6 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/molecules/ContactModalForm/ContactModalForm.tsx:182]: Owner is read only here and submit validates the id, not the object. Keeping.

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx (hunk 7 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: Gone. That effect no longer exists after the deal panel split.

File: frontend/src/community/crm/v2/components/molecules/AddContactModalContent/AddContactModalContent.tsx (hunk 8 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/molecules/AddContactModalContent/AddContactModalContent.tsx:47]: Fixed. Seeding the owner without reinitialising the form.

File: frontend/src/community/crm/v2/components/molecules/EditContactModalContent/EditContactModalContent.tsx (hunk 9 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/molecules/EditContactModalContent/EditContactModalContent.tsx:122]: Edit flow has no add company option. Keeping.

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx (hunk 10 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: The guard is back on develop. Gone from here.

File: frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx (hunk 11 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/organisms/DealSidePanelV2/DealSidePanelV2.tsx]: Gone with the deal panel split.

File: frontend/src/community/crm/v2/components/molecules/EditableContactCompanyField/EditableContactCompanyField.tsx (hunk 12 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/components/molecules/EditableContactCompanyField/EditableContactCompanyField.tsx]: Fixed. Merged.

File: frontend/src/community/crm/v2/hooks/useContactEmailCheck.ts (hunk 13 above)
Comment (dulsara-manakal) [reply frontend/src/community/crm/v2/hooks/useContactEmailCheck.ts]: Moved into the form that uses it. Only one caller.

**Review by akilarootcode — DISMISSED** [review]

**Review by hasalarootcode — DISMISSED** [review]

**Review by ThinuwanW — APPROVED** [review]
